Mage.Sets/src/mage/cards/h/HinataDawnCrowned.java

```diff
@@ -0,0 +1,129 @@
+package mage.cards.h;
+
+import mage.MageInt;
+import mage.abilities.Ability;
+import mage.abilities.SpellAbility;
+import mage.abilities.common.SimpleStaticAbility;
+import mage.abilities.effects.common.cost.CostModificationEffectImpl;
+import mage.abilities.keyword.FlyingAbility;
+import mage.abilities.keyword.TrampleAbility;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.constants.*;
+import mage.game.Game;
+import mage.util.CardUtil;
+
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
+
+/**
+ * @author TheElk801
+ */
+public final class HinataDawnCrowned extends CardImpl {
+
+    public HinataDawnCrowned(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{1}{U}{R}{W}");
+
+        this.addSuperType(SuperType.LEGENDARY);
+        this.subtype.add(SubType.KIRIN);
+        this.subtype.add(SubType.SPIRIT);
+        this.power = new MageInt(4);
+        this.toughness = new MageInt(4);
+
+        // Flying
+        this.addAbility(FlyingAbility.getInstance());
+
+        // Trample
+        this.addAbility(TrampleAbility.getInstance());
+
+        // Spells you cast cost {1} less to cast for each target.
+        this.addAbility(new SimpleStaticAbility(new HinataDawnCrownedOwnEffect()));
+
+        // Spells your opponents cast cost {1} more to cast for each target
+        this.addAbility(new SimpleStaticAbility(new HinataDawnCrownedOpponentsEffect()));
+    }
+
+    private HinataDawnCrowned(final HinataDawnCrowned card) {
+        super(card);
+    }
+
+    @Override
+    public HinataDawnCrowned copy() {
+        return new HinataDawnCrowned(this);
+    }
+}
+
+class HinataDawnCrownedOwnEffect extends CostModificationEffectImpl {
+
+    HinataDawnCrownedOwnEffect() {
+        super(Duration.WhileOnBattlefield, Outcome.Neutral, CostModificationType.REDUCE_COST);
+        staticText = "Spells you cast cost {1} less to cast for each target";
+    }
+
+    private HinataDawnCrownedOwnEffect(HinataDawnCrownedOwnEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public HinataDawnCrownedOwnEffect copy() {
+        return new HinataDawnCrownedOwnEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source, Ability abilityToModify) {
+        Set<UUID> targets = null;
+        if (game.inCheckPlayableState()) {
+            targets = CardUtil.getAllPossibleTargets(abilityToModify, game);
+        } else {
+            targets = CardUtil.getAllSelectedTargets(abilityToModify, game);
+        }
+        for (int i = 0; i < targets.stream().count(); i++) {
+            CardUtil.reduceCost(abilityToModify, 1);
```

Comment:
i think this can be optimized a bit by replacing the loop with:

CardUtil.reduceCost(abilityToModify, targets.stream().count());

Assuming that method works as I think it does. Let me know if that doesn't work, otherwise I will make that change

Reply:
targets.size() is even better (constant time)

Reply:
you are correct. I made that change in this [PR](https://github.com/magefree/mage/pull/8657)